scipy/ndimage/src/_rank_filter_1d.cpp

```diff
@@ -0,0 +1,309 @@
+    }
+  } else // new item is at rank
+  {
+    if (maxSortUp(data, m, -1)) {
+      maxSortDown(data, m, -1);
+    }
+    if (minSortUp(data, m, 1)) {
+      minSortDown(data, m, 1);
+    }
+  }
+}
+
+template <typename T>
+void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
+                  int mode, T cval, int origin) {
+  int i, arr_len_thresh, lim = (win_len - 1) / 2 - origin;
+  int lim2 = arr_len - lim;
+  Mediator *m = MediatorNew(win_len, rank);
+  T *data = new T[win_len];
+
+  switch (mode) {
+  case REFLECT:
+    for (i = win_len - lim - 1; i > -1; i--) {
+      MediatorInsert(data, m, in_arr[i]);
+    }
+    break;
+  case CONSTANT:
+    for (i = win_len - lim; i > 0; i--) {
+      MediatorInsert(data, m, cval);
+    }
+    break;
+  case NEAREST:
+    for (i = win_len - lim; i > 0; i--) {
+      MediatorInsert(data, m, in_arr[0]);
+    }
+    break;
+  case MIRROR:
+    for (i = win_len - lim; i > 0; i--) {
+      MediatorInsert(data, m, in_arr[i]);
+    }
+    break;
+  case WRAP:
+    for (i = arr_len - lim - 1 - 2 * origin; i < arr_len; i++) {
+      MediatorInsert(data, m, in_arr[i]);
+    }
+    break;
+  }
+
+  for (i = 0; i < lim; i++) {
+    MediatorInsert(data, m, in_arr[i]);
+  }
+  for (i = lim; i < arr_len; i++) {
+    MediatorInsert(data, m, in_arr[i]);
+    out_arr[i - lim] = data[m->heap[0]];
+  }
+  switch (mode) {
+  case REFLECT:
+    arr_len_thresh = arr_len - 1;
+    for (i = 0; i < lim; i++) {
+      MediatorInsert(data, m, in_arr[arr_len_thresh - i]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case CONSTANT:
+    for (i = 0; i < lim; i++) {
+      MediatorInsert(data, m, cval);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case NEAREST:
+    arr_len_thresh = arr_len - 1;
+    for (i = 0; i < lim; i++) {
+      MediatorInsert(data, m, in_arr[arr_len_thresh]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case MIRROR:
+    arr_len_thresh = arr_len - 2;
+    for (i = 0; i < lim + 1; i++) {
+      MediatorInsert(data, m, in_arr[arr_len_thresh - i]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case WRAP:
+    for (i = 0; i < win_len; i++) {
+      MediatorInsert(data, m, in_arr[i]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  }
+
+  m->heap -= rank;
+  delete[] m->heap;
+  m->heap = nullptr;
+  delete[] m->pos;
+  m->pos = nullptr;
+  delete m;
+  m = nullptr;
+  delete[] data;
+  data = nullptr;
+}
+
+// Python wrapper for rank_filter
+static PyObject *rank_filter(PyObject *self, PyObject *args) {
+  PyObject *in_arr_obj, *out_arr_obj, *cval_obj;
+  int rank, arr_len, win_len, mode, origin, type;
+  if (!PyArg_ParseTuple(args, "OiiOiOi", &in_arr_obj, &rank, &win_len,
+                        &out_arr_obj, &mode, &cval_obj, &origin)) {
+    return NULL;
+  }
+  PyArrayObject *in_arr = (PyArrayObject *)PyArray_FROM_OTF(
+      in_arr_obj, NPY_NOTYPE, NPY_ARRAY_INOUT_ARRAY2);
```

Comment:
Am pretty sure the problem is here. While I don't know exactly what `PyArray_FROM_OTF` is doing, its documentation mentions that F is for arrays flags, and the problem is with flags. So here we are. 

Now, if you make sure that `input` is indeed a numpy array on the python side, AFAIU you can just cast the `PyObject *` pointer from `PyArg_ParseTuple` to a `PyArrayObject *` pointer. This will be a borrowed reference, so no need to INCREF, can just pass its data to `_rank_filter` like you do below.